tools/cjxl_ng_main.cc

```diff
@@ -235,37 +216,23 @@ DEFINE_int32(photon_noise, 0,
 DEFINE_int32(codestream_level, 5, "The codestream level. Either `5` or `10`.");
```

Comment:
It might make sense to have a default of -1 here, meaning "let the encoder set the codestream level to the minimum needed to represent the input". Otherwise you have to manually set the level to 10 when you're e.g. trying to losslessly encode a 16-bit png or a 32-bit pfm, or an image that is more than 268 megapixels, since level 5 encoding will fail for those.

Reply:
Do we actually know what the "minimum needed" is before we have seen every frame?

Reply:
No, we don't, at least not in the case of animations, since some restrictions are things like the minimum nonzero frame duration and the maximum number of pixels in consecutive zero-duration frames.

But most of the level restrictions apply to things that can be known early, like bitdepth, number of extra channels, and image dimensions.

Reply:
So... the idea here is that if we see anything in the parameters that force level 10, we set level 10 straightaway, and if it can be determined only while processing frames, and the user did not set level 10, they get a failure message?

Sounds like a reasonable for-convenience idea, overall, but the API documentation should have a list of things that automatically trigger level 10. Not being too familiar with the internals here, I would struggle to come up with a complete list of criteria, and in the interest of having behavior documented, cjxl_ng users should be able to reason about when this automatic switch to level 10 happens.

Reply:
For JPEG recompression, since we only do 8-bit JPEGs without CMYK, level 5 should usually be enough. The exception is if it's a huge JPEG with more than 1 << 28 pixels or one with a very large ICC profile.

In terms of API, I think we need to clarify how we want the API to work, and then make sure it works like that and properly document it. I see three options:

- Application is responsible for setting the codestream level _before_ setting basic info, and setting basic info fails when trying to set things that are not compatible with the codestream level. In this case, GetRequiredCodestreamLevel is not useful and should get deprecated. This is quite inconvenient because it means applications have to know what level to set.
- Application is responsible for setting the codestream level _after_ setting basic info (but before encoding the first frame), and can use GetRequiredCodestreamLevel to set it based on what is implied by basic info. This is somewhat error-prone since applications will likely forget to call GetRequiredCodestreamLevel and set the level accordingly, which will usually be fine (since many images can be done in the default level 5), but will cause unexpected failure when e.g. the dimensions are too large or when trying to encode a 16-bit image.
- Application can optionally manually set the codestream level, but if it doesn't, the encoder automatically picks the lowest level that is compatible with what is known when it has to decide what to write as the codestream level. So if it wasn't manually set, it chooses the lowest level possible, and if it was manually set, it uses that level and returns error if the level turns out to be too low. In this case, GetRequiredCodestreamLevel is not needed. This is probably the most convenient way to do it. It also has the advantage that AddJPEGFrame can still set the level to 10 in case of a very large JPEG.

Reply:
> * but will cause unexpected failure when e.g. the dimensions are too large or when trying to encode a 16-bit image.

Is level 10 really necessary for 16bit images?

According https://github.com/libjxl/libjxl/blob/eb152f5259df9af806c5b47b192072444b9dc41b/lib/include/jxl/encode.h#L943 level 5 has supports _maximum 16-bit color channel depth_.

Reply:
Level 10 is necessary to do _lossless_ 16-bit.

Technically, what makes the difference is if `int16_t` buffers can be used for the modular parts or not. When doing lossy VarDCT, this will be fine, since modular is then only used for quantized DC and control fields, which are all things that fit in `int16_t`. But when doing lossless Modular with uint16 input, `int16_t` is not enough. The current encoder will signal that `int16_t` is not enough whenever trying to do lossless with more than 12-bit input (it puts the boundary there so it still has spare bits for doing RCT and/or Squeeze), and that means level 5 is not enough.

Note that this is unrelated to how buffers are passed in the API, e.g. you can still get/set buffers using UINT16 or FLOAT32 sample types regardless of the codestream level.

Reply:
This is important information.

16bit lossless always worked in 0.6.1

But now in git master I have to set codestream level to 10.

Reply:
In 0.6.1, the encoder is producing a codestream that is not properly marked as being level 10. The current decoder doesn't care about it, but it's kind of an encoder bug. But I do think that it would be a good idea to implement the third option I proposed, i.e. the encoder automatically sets the level to 10 when needed.